Mathlib/FieldTheory/Finite/Basic.lean

```diff
@@ -179,6 +179,33 @@ theorem forall_pow_eq_one_iff (i : ℕ) : (∀ x : Kˣ, x ^ i = 1) ↔ q - 1 ∣
       rw [← pow_mul, mul_comm, pow_mul, h, one_pow]
 #align finite_field.forall_pow_eq_one_iff FiniteField.forall_pow_eq_one_iff
 
+theorem cast_subgroup_of_units_card_ne_zero {F : Type} [Field F]
+    (G : Subgroup Fˣ) [Fintype G] : (Fintype.card G : F) ≠ 0 := by
+  let n := Fintype.card G
+  intro nzero
+  have ⟨p, char_p⟩ := CharP.exists F
+  have hd : p ∣ n := (CharP.cast_eq_zero_iff F p n).mp nzero
+  cases CharP.char_is_prime_or_zero F p with
+  | inr pzero =>
+    exact ne_zero_of_lt (Fintype.card_pos) $ Nat.eq_zero_of_zero_dvd $ pzero ▸ hd
+  | inl pprime =>
+    have fact_pprime := Fact.mk pprime
+    -- G has an element x of order p by Cauchy's theorem
+    have ⟨x, hx⟩ := exists_prime_orderOf_dvd_card p hd
+    -- F has an element u (= ↑↑x) of order p
+    let u := ((x : Fˣ) : F)
+    have hu : orderOf u = p := by rwa [orderOf_units, orderOf_subgroup]
+    -- u ^ p = 1 implies (u - 1) ^ p = 0 and hence u = 1 ...
+    have h : u = 1 := by
+      rw [← sub_left_inj, sub_self 1]
+      apply pow_eq_zero (n := p)
+      rw [sub_pow_char (R := F) (p := p) u 1, one_pow, ← hu, pow_orderOf_eq_one]
+      exact sub_self 1
+    -- ... meaning x didn't have order p after all, contradiction
+    apply ne_of_lt $ Nat.Prime.one_lt pprime
+    rw [← hu, h]
+    exact orderOf_one.symm
```

Comment:
A few golfs, and actually this holds for a not-necessarily commutative ring with no zero divisors & nontrivial, and I don't see why it would hold in more generality than this.

`<|` is preferred in Lean4 to `$`, as far as I can see. 

```suggestion
theorem cast_subgroup_of_units_card_ne_zero {F : Type} [Ring F] [NoZeroDivisors F] [Nontrivial F]
    (G : Subgroup Fˣ) [Fintype G] : (Fintype.card G : F) ≠ 0 := by
  let n := Fintype.card G
  intro nzero
  have ⟨p, char_p⟩ := CharP.exists F
  have hd : p ∣ n := (CharP.cast_eq_zero_iff F p n).mp nzero
  cases CharP.char_is_prime_or_zero F p with
  | inr pzero =>
    exact (Fintype.card_pos).ne' <| Nat.eq_zero_of_zero_dvd <| pzero ▸ hd
  | inl pprime =>
    have fact_pprime := Fact.mk pprime
    -- G has an element x of order p by Cauchy's theorem
    have ⟨x, hx⟩ := exists_prime_orderOf_dvd_card p hd
    -- F has an element u (= ↑↑x) of order p
    let u := ((x : Fˣ) : F)
    have hu : orderOf u = p := by rwa [orderOf_units, orderOf_subgroup]
    -- u ^ p = 1 implies (u - 1) ^ p = 0 and hence u = 1 ...
    have h : u = 1 := by
      rw [← sub_left_inj, sub_self 1]
      apply pow_eq_zero (n := p)
      rw [sub_pow_char_of_commute, one_pow, ← hu, pow_orderOf_eq_one, sub_self]
      exact Commute.one_right u
    -- ... meaning x didn't have order p after all, contradiction
    apply pprime.one_lt.ne
    rw [← hu, h, orderOf_one]
```

Reply:
Nice catch on the generality!